File: python/kalman/ekf.py
"""Extended Kalman filter (numpy port of the C ``kf_ekf_*`` API).

Nonlinear state-transition and measurement models, linearised with the user's
Jacobians:

    x(k+1) = f(x, u, dt) + w,    z = h(x) + v
    F = df/dx,  H = dh/dx   (evaluated at the current estimate)
"""

from __future__ import annotations

import numpy as np

from .matrix import symmetrize


class ExtendedKalmanFilter:
    """Extended Kalman filter.

    Parameters
    ----------
    n, m : int
        State / measurement dimension.
    f : callable(x, u, dt) -> x_out
        State-transition function.
    F_jac : callable(x, u, dt) -> F (n, n)
        Jacobian of f w.r.t. the state.
    h : callable(x) -> z_out
        Measurement function.
    H_jac : callable(x) -> H (m, n)
        Jacobian of h w.r.t. the state.
    """

    def __init__(self, n, m, f=None, F_jac=None, h=None, H_jac=None):
        self.n = int(n)
        self.m = int(m)
        self.x = np.zeros(self.n)
        self.P = np.eye(self.n)
        self.Q = np.zeros((self.n, self.n))
        self.R = np.zeros((self.m, self.m))

        self.f = f
        self.F_jac = F_jac
        self.h = h
        self.H_jac = H_jac

        self._y = np.zeros(self.m)

    def set_models(self, f, F_jac, h, H_jac) -> None:
        """Register the four nonlinear model callbacks."""
        self.f, self.F_jac, self.h, self.H_jac = f, F_jac, h, H_jac

    # --- state / covariance / noise (same semantics as KalmanFilter) ---
    def set_state(self, x) -> None:
        self.x[:] = np.asarray(x, dtype=float)

    def get_state(self) -> np.ndarray:
        return self.x.copy()

    def set_covariance(self, P) -> None:
        self.P[:] = np.asarray(P, dtype=float)

    def set_covariance_diagonal(self, diag) -> None:
        self.P[:] = np.diag(np.asarray(diag, dtype=float))

    def set_covariance_scalar(self, p) -> None:
        self.P[:] = float(p) * np.eye(self.n)

    def get_covariance(self) -> np.ndarray:
        return self.P.copy()

    def set_process_noise(self, Q) -> None:
        self.Q[:] = np.asarray(Q, dtype=float)

    def set_process_noise_scalar(self, q) -> None:
        self.Q[:] = float(q) * np.eye(self.n)

    def set_measurement_noise(self, R) -> None:
        self.R[:] = np.asarray(R, dtype=float)

    def set_measurement_noise_scalar(self, r) -> None:
        self.R[:] = float(r) * np.eye(self.m)

    # --- filtering ---
    def predict(self, u=None, dt=0.0) -> None:
        if self.f is None or self.F_jac is None:
            raise RuntimeError("models not set; call set_models() first")
        F = np.asarray(self.F_jac(self.x, u, dt), dtype=float)
        if F.shape != (self.n, self.n):
            raise ValueError("F_jac must return (%d,%d)" % (self.n, self.n))
        x_new = np.asarray(self.f(self.x, u, dt), dtype=float)
        self.x[:] = x_new
        self.P[:] = symmetrize(F @ self.P @ F.T + self.Q)

    def update(self, z) -> None:
        if self.h is None or self.H_jac is None:
            raise RuntimeError("models not set; call set_models() first")
        z = np.asarray(z, dtype=float)
        H = np.asarray(self.H_jac(self.x), dtype=float)
        if H.shape != (self.m, self.n):
            raise ValueError("H_jac must return (%d,%d)" % (self.m, self.n))
        z_hat = np.asarray(self.h(self.x), dtype=float)
        y = z - z_hat
        self._y[:] = y

        P, R = self.P, self.R
        S = H @ P @ H.T + R
        K = np.linalg.solve(S, (P @ H.T).T).T
        self.x = self.x + K @ y
        IKH = np.eye(self.n) - K @ H
        self.P[:] = symmetrize(IKH @ P @ IKH.T + K @ R @ K.T)

File: matlab/+kalman/EKF.m
classdef EKF < handle
%EKF Extended Kalman filter (MATLAB port of the C kf_ekf_* API).
%
%   Nonlinear models, linearised with the user's Jacobians:
%     x(k+1) = f(x, u, dt) + w,   z = h(x) + v
%     F = df/dx,  H = dh/dx  (at the current estimate)

    properties
        n (1,1) double = 0
        m (1,1) double = 0
        x  double
        P  double
        Q  double
        R  double
        f     function_handle   % f(x, u, dt) -> x_out
        F_jac function_handle   % F_jac(x, u, dt) -> F (n x n)
        h     function_handle   % h(x) -> z_out
        H_jac function_handle   % H_jac(x) -> H (m x n)
    end

    methods
        function obj = EKF(n, m, f, F_jac, h, H_jac)
            if nargin < 5, H_jac = []; end
            if nargin < 4, h = []; end
            if nargin < 3, F_jac = []; end
            if nargin < 2, f = []; end
            obj.n = n;
            obj.m = m;
            obj.x = zeros(n, 1);
            obj.P = eye(n);
            obj.Q = zeros(n);
            obj.R = zeros(m);
            obj.f = f; obj.F_jac = F_jac; obj.h = h; obj.H_jac = H_jac;
        end

        function set_models(obj, f, F_jac, h, H_jac)
            obj.f = f; obj.F_jac = F_jac; obj.h = h; obj.H_jac = H_jac;
        end

        function set_state(obj, x), obj.x = double(x(:)); end
        function x = get_state(obj), x = obj.x; end
        function set_covariance(obj, P), obj.P = double(P); end
        function set_covariance_diagonal(obj, d), obj.P = diag(double(d(:))); end
        function set_covariance_scalar(obj, p), obj.P = p * eye(obj.n); end
        function P = get_covariance(obj), P = obj.P; end
        function set_process_noise(obj, Q), obj.Q = double(Q); end
        function set_process_noise_scalar(obj, q), obj.Q = q * eye(obj.n); end
        function set_measurement_noise(obj, R), obj.R = double(R); end
        function set_measurement_noise_scalar(obj, r), obj.R = r * eye(obj.m); end

        function predict(obj, u, dt)
            if nargin < 3, dt = 0; end
            if nargin < 2, u = []; end
            if isempty(obj.f) || isempty(obj.F_jac)
                error('kalman:EKF:noModel', 'call set_models() first');
            end
            F = double(obj.F_jac(obj.x, u, dt));
            obj.x = double(obj.f(obj.x, u, dt));
            obj.P = F * obj.P * F' + obj.Q;
            obj.P = (obj.P + obj.P') / 2;
        end

        function update(obj, z)
            z = double(z(:));
            if isempty(obj.h) || isempty(obj.H_jac)
                error('kalman:EKF:noModel', 'call set_models() first');
            end
            H = double(obj.H_jac(obj.x));
            y = z - double(obj.h(obj.x));
            S = H * obj.P * H' + obj.R;
            K = (S \ (obj.P * H')')';
            obj.x = obj.x + K * y;
            IKH = eye(obj.n) - K * H;
            obj.P = IKH * obj.P * IKH' + K * obj.R * K';
            obj.P = (obj.P + obj.P') / 2;
        end
    end
end
